Prelude/NonEmpty.swift

```diff
@@ -0,0 +1,100 @@
+public struct NonEmpty<Collection: Swift.Collection> {
+  public typealias Element = Collection.Iterator.Element
+  public let head: Collection.Iterator.Element
+  public let tail: Collection
+}
+
+infix operator >|: AdditionPrecedence
+
+public func >| <C: Collection>(head: C.Iterator.Element, tail: C) -> NonEmpty<C> {
+  return NonEmpty<C>(head: head, tail: tail)
+}
+
+extension NonEmpty {
+  public var count: Collection.IndexDistance {
+    return tail.count.advanced(by: 1)
+  }
+
+  public var first: Element {
+    return head
+  }
+}
+
+extension NonEmpty where Collection: RandomAccessCollection {
+  public var last: Element {
+    return tail.last ?? head
+  }
+}
+
+extension Array {
+  public init<C: Collection>(_ nonEmpty: NonEmpty<C>) where C.Iterator.Element == Element {
+    self = [nonEmpty.head] + Array(nonEmpty.tail)
+  }
+}
+
+// MARK: Array
+
+public typealias NonEmptyArray<Element> = NonEmpty<[Element]>
+
+public func == <T>(lhs: NonEmptyArray<T>, rhs: NonEmptyArray<T>) -> Bool where T: Equatable {
+  return lhs.head == rhs.head && lhs.tail == rhs.tail
+}
+
+public func != <T>(lhs: NonEmptyArray<T>, rhs: NonEmptyArray<T>) -> Bool where T: Equatable {
+  return lhs.head != rhs.head && lhs.tail != rhs.tail
+}
+
+public protocol ArrayType {
```

Comment:
Using these protocols to allow for the variadic initializers, _e.g._ `NonEmptyArray(1, 2, 3)` (which double as convenience singleton initializers: `NonEmptyArray(1)` _vs._ `1 >| []`). If we can do without 'em, please let me know!